File: libs/language/src/lib/provider/AbstractNodeProvider.ts
import { NodeProvider } from '../NodeProvider';
import { BlueNode, NodeDeserializer } from '../model';
import { NodeContentHandler } from './NodeContentHandler';
import { JsonBlueValue } from '../../schema';

export abstract class AbstractNodeProvider extends NodeProvider {
  /**
   * Fetches raw content by Blue ID
   * @param baseBlueId - The base Blue ID (without # suffix)
   * @returns The raw content as JSON-like object, or null if not found
   */
  protected abstract fetchContentByBlueId(
    baseBlueId: string,
  ): JsonBlueValue | null;

  override fetchByBlueId(blueId: string): BlueNode[] | null {
    const baseBlueId = blueId.split('#')[0];
    const content = this.fetchContentByBlueId(baseBlueId);

    if (content === null || content === undefined) {
      return null;
    }

    const isMultipleDocuments = Array.isArray(content) && content.length > 1;
    const resolvedContent = NodeContentHandler.resolveThisReferences(
      content,
      baseBlueId,
      isMultipleDocuments,
    );

    if (blueId.includes('#')) {
      const parts = blueId.split('#');
      if (parts.length > 1) {
        const index = parseInt(parts[1]);
        if (Array.isArray(resolvedContent) && index < resolvedContent.length) {
          const item = resolvedContent[index];
          const node = NodeDeserializer.deserialize(item);

          node.setBlueId(blueId);
          return [node];
        } else if (index === 0) {
          const node = NodeDeserializer.deserialize(resolvedContent);
          node.setBlueId(blueId);
          return [node];
        } else {
          return null;
        }
      }
    }

    if (Array.isArray(resolvedContent)) {
      return resolvedContent.map((item) => {
        const node = NodeDeserializer.deserialize(item);
        return node;
      });
    } else {
      const node = NodeDeserializer.deserialize(resolvedContent);
      node.setBlueId(baseBlueId);
      return [node];
    }
  }
}

File: references/java-sdk/src/main/java/blue/language/provider/AbstractNodeProvider.java
package blue.language.provider;

import blue.language.NodeProvider;
import blue.language.model.Node;
import com.fasterxml.jackson.databind.JsonNode;

import java.util.Collections;
import java.util.List;
import java.util.stream.Collectors;
import java.util.stream.IntStream;

import static blue.language.utils.UncheckedObjectMapper.JSON_MAPPER;

public abstract class AbstractNodeProvider implements NodeProvider {

    @Override
    public List<Node> fetchByBlueId(String blueId) {
        final String baseBlueId = blueId.split("#")[0];
        final JsonNode content = fetchContentByBlueId(baseBlueId);
        if (content == null) {
            return null;
        }

        boolean isMultipleDocuments = content.isArray() && content.size() > 1;
        final JsonNode resolvedContent = NodeContentHandler.resolveThisReferences(content, baseBlueId, isMultipleDocuments);

        if (blueId.contains("#")) {
            String[] parts = blueId.split("#");
            if (parts.length > 1) {
                int index = Integer.parseInt(parts[1]);
                if (resolvedContent.isArray() && index < resolvedContent.size()) {
                    JsonNode item = resolvedContent.get(index);
                    Node node = JSON_MAPPER.convertValue(item, Node.class);
                    return Collections.singletonList(node.blueId(blueId));
                } else if (index == 0) {
                    Node node = JSON_MAPPER.convertValue(resolvedContent, Node.class);
                    return Collections.singletonList(node.blueId(blueId));
                } else {
                    return null;
                }
            }
        }

        if (resolvedContent.isArray()) {
            return IntStream.range(0, resolvedContent.size())
                    .mapToObj(i -> JSON_MAPPER.convertValue(resolvedContent.get(i), Node.class))
                    .collect(Collectors.toList());
        } else {
            Node node = JSON_MAPPER.convertValue(resolvedContent, Node.class);
            return Collections.singletonList(node.blueId(baseBlueId));
        }
    }

    protected abstract JsonNode fetchContentByBlueId(String baseBlueId);
}